### ISSUE TITLE
NX Parallel doesn't properly sequence the dependent packages

### ISSUE THREAD

**Issue by brsanthu** [nrwl/nx#28599]
### Current Behavior

We have a monorepo with about 15 packages with inter-dependencies but no circular dependencies. When we run nx to process in parallel, it always fails as it tries to compile dependent package before compiling its dependency. Same command works when we run `--parallel 1`.

### Expected Behavior

Nx should take dependency into consideration before processing dependent packages

### Steps to Reproduce

1. Have a repo with multiple packages with dependencies
2. Try to process with concurrency higher than 1. For ex., `--parallel 5`

### Nx Report
```
❯ nx report

 NX   Report complete - copy this into the issue template

Node           : 20.10.0
OS             : darwin-arm64
Native Target  : aarch64-macos
yarn           : 4.4.1

nx (global)    : 20.0.5
nx             : 20.0.5
@nx/js         : 20.0.5
@nx/workspace  : 20.0.5
@nx/devkit     : 20.0.5
@nrwl/tao      : 19.8.4
typescript     : 5.5.4
---------------------------------------
The following packages should match the installed version of nx
  - @nrwl/tao@19.8.4

To fix this, run `nx migrate nx@20.0.5`
```

### Failure Logs

This is with [nx run-many -t tsc --parallel 5]. Notice that @company/js-csv depends on @company/js-utils but js-utils is not processed first.
```
❯ yarn tsc

 NX   Running target tsc for 32 projects

   ✖  nx run @company/js-csv:tsc
      src/csvUtils.ts(3,32): error TS2307: Cannot find module '@company/js-utils' or its corresponding type declarations.
      src/models.ts(1,31): error TS2307: Cannot find module '@company/js-utils' or its corresponding type declarations.
```

This is with [nx run-many -t tsc --parallel 1]. Notice that js-utils is compiled first as that is base for all other packages.
```
❯ yarn tsc

   ✔  nx run @company/js-utils:tsc (2s)
   ✔  nx run @company/js-csv:tsc (2s)
```

### Operating System

- [ ] Linux
- [x] macOS
- [ ] Windows
- [ ] Other (Please specify)

**Comment by Coly010** [comment]
@brsanthu Nx _does_ take dependency order into account. If you're noticing that it is not happening, it is either a bug, or a configuration issue. 

Is there any way you could provide a repo that reproduces this issue so that we can investigate it more thoroughly? It's hard to reproduce as we have a bias of experience in setting up an Nx workspace wherein the Project Graph would have the correct edges.

**Comment by phillipCouto** [comment]
We are also experiencing a similar issue, we are only seeing this when we use `affected` or `run-many`. When we run the individual task it will execute the dependent tasks in the correct order. Also when we look at the graph we are seeing the tasks identified in the correct order.

This issue popped up when we went from nx 19.4.x to 19.5.x.

The graph is like so:

```
libs/backend/models:migrate-db -> libs/backend/models:prebuild -> libs/backend/requestcache:audit
```

The `libs/backend/models` package does have many other projects depend on it and it does execute correctly at the beginning but nx will incorrectly schedule a dependent task to run in parallel with `libs/backend/models:prebuild`. This scheduled task will fail but then the remaining dependent tasks will completely successfully.

I ran an additional test were I artificially added a long delay to `libs/backend/models:prebuild` and nx did wait correctly to schedule any new tasks that depend on it until `libs/backend/models:prebuild` completes.

So far with my testing it seems nx is incorrectly scheduling dependent tasks if there is parallel capacity. Once `libs/backend/models:prebuild` is started then nx will wait.

I will follow-up with a repro of the issue tonight if no one else posts one but I wanted to document my observations thus far.

**Comment by phillipCouto** [comment]
So it seems the issue I was experiencing in 19.5.x was resolved in 19.6.x. Funny enough there is a similar issue that was introduced in 19.8.4 that is still present even in 20.0.5.

I have created a project with instructions on how to reproduce. This was done on a macbook but I also tested it on a fedora core VM and the behaviour is the same. Due to the commands I used to simulate the tasks it may not work on windows.

https://github.com/phillipCouto/nx-task-ordering-issue-19.8.4
